rts/motoko-rts/src/gc/generational/remembered_set.rs

```diff
@@ -0,0 +1,282 @@
+//! Remembered set.
+//! Serves for recording pointer locations trapped by the write barrier.
+//!
+//! Hash-set implementation. Linked-list collision handling.
+//!
+//! Hash function = (ptr / WORD_SIZE) % TABLE_SIZE
+//!
+//! Hash-table (length N):
+//! ----------------------------------
+//! | entry[0]   | collision_ptr[0]   |
+//! ----------------------------------
+//! | entry[1]   | collision_ptr[1]   |
+//! ----------------------------------
+//! |  ...                           |
+//! ----------------------------------
+//! | entry[N-1] | collision_ptr[N-1] |
+//! ----------------------------------
+//!
+//! Per collision a new linked list node is appended:
+//!
+//!                               Collision node
+//!                        ------------------------------
+//! prev_collision_ptr --> | entry | next_collision_ptr |
+//!                        ------------------------------
+//!
+//! Amortized hash-table growth when exceeding a defined threshold.
+//!
+//! Growth factor 2 for faster bitwise modulo calculation.
+//!
+//! NOTE: Remembered set structure is not recorded by write barriers
+//! as it is discarded by each GC run.
+//!
+//! NOTE: The table must be blobs, as their entries must not be
+//! analyzed by the GC.
+
+use core::mem::size_of;
+use core::ptr::null_mut;
+
+use crate::constants::WORD_SIZE;
+use crate::memory::{alloc_blob, Memory};
+use crate::types::{object_size, Blob, Bytes, Value};
+
+pub struct RememberedSet {
+    hash_table: *mut Blob,
+    count: u32,        // contained entries
+    fast_cache: Value, // optimization, last inserted value
+}
+
+#[repr(C)]
+struct HashEntry {
+    pub value: Value,
+    pub next_collision_ptr: *mut CollisionNode,
+}
+
+#[repr(C)]
+struct CollisionNode {
+    pub header: Blob,
+    pub entry: HashEntry,
+}
+
+pub struct RememberedSetIterator {
+    hash_table: *mut Blob,
+    hash_index: u32,
+    current_entry: *mut HashEntry,
+}
+
+pub const INITIAL_TABLE_LENGTH: u32 = 1024;
+const GROWTH_FACTOR: u32 = 2;
+pub const OCCUPATION_THRESHOLD_PERCENT: u32 = 65;
+
+impl RememberedSet {
+    pub unsafe fn new<M: Memory>(mem: &mut M) -> RememberedSet {
+        let hash_table = new_table(mem, INITIAL_TABLE_LENGTH);
+        RememberedSet {
+            hash_table,
+            count: 0,
+            fast_cache: null_ptr_value(),
+        }
+    }
+
+    pub unsafe fn insert<M: Memory>(&mut self, mem: &mut M, value: Value) {
+        if is_null_ptr_value(value) || value.get_raw() == self.fast_cache.get_raw() {
+            return;
+        }
+        let index = self.hash_index(value);
```

Comment:
Shouldn't you update fast_cache here? Or is it set elsewhere?

Reply:
It looks to me like fast_cache is only ever set to null_ptr_value() (at line 77)

Reply:
Thanks a lot again. Indeed, never updated. I will fix this asap. Maybe, performance improves. 
(Btw: This bug probably happened while changing from remembered log to remembered set.)